src/common/meta/src/kv_backend.rs

```diff
@@ -13,68 +13,133 @@
 // limitations under the License.
 
 pub mod memory;
+pub mod txn;
 
 use std::any::Any;
-use std::pin::Pin;
 use std::sync::Arc;
 
 use async_trait::async_trait;
 use common_error::ext::ErrorExt;
-use futures::{Stream, StreamExt};
+pub use txn::TxnService;
 
 use crate::error::Error;
+use crate::rpc::store::{
+    BatchDeleteRequest, BatchDeleteResponse, BatchGetRequest, BatchGetResponse, BatchPutRequest,
+    BatchPutResponse, CompareAndPutRequest, CompareAndPutResponse, DeleteRangeRequest,
+    DeleteRangeResponse, MoveValueRequest, MoveValueResponse, PutRequest, PutResponse,
+    RangeRequest, RangeResponse,
+};
+use crate::rpc::KeyValue;
 
-#[derive(Debug, Clone, PartialEq)]
-pub struct Kv(pub Vec<u8>, pub Vec<u8>);
+pub type KvBackendRef = Arc<dyn KvBackend<Error = Error> + Send + Sync>;
 
-pub type ValueIter<'a, E> = Pin<Box<dyn Stream<Item = Result<Kv, E>> + Send + 'a>>;
+#[async_trait]
+pub trait KvBackend: TxnService
+where
+    Self::Error: ErrorExt + Send + Sync,
+{
+    fn name(&self) -> &'static str;
+
+    async fn range(&self, req: RangeRequest) -> Result<RangeResponse, Self::Error>;
+
+    async fn put(&self, req: PutRequest) -> Result<PutResponse, Self::Error>;
+
+    /// Default "batch_put" is backed by iterating all kvs on "put".
+    async fn batch_put(&self, req: BatchPutRequest) -> Result<BatchPutResponse, Self::Error> {
```

Comment:
This default implementation does not guarantee atomicity, which violates the semantic of `batch_put` operation. Maybe we'd better remove the default impl.